File: sdks/node/src/distributed-fuzz.js
import { AsyncLocalStorage } from "node:async_hooks";
import { createHash, createPublicKey, verify } from "node:crypto";

import { canonicalBytes } from "./encoding.js";

export const FUZZ_CONTEXT_HTTP_HEADER = "ReproIt-Fuzz-Context";
export const FUZZ_PARENT_HTTP_HEADER = "ReproIt-Parent-Operation";
export const FUZZ_CONTEXT_QUEUE_METADATA = "reproit.fuzz.context";
export const FUZZ_PARENT_QUEUE_METADATA = "reproit.parent.operation";

const MAX_CONTEXT_BYTES = 4_096;
const UUID7 = "[0-9a-f]{8}-[0-9a-f]{4}-7[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}";
const CAMPAIGN_ID = new RegExp(`^fc_${UUID7}$`);
const CASE_ID = new RegExp(`^case_${UUID7}$`);
const PROJECT_ID = new RegExp(`^prj_${UUID7}$`);
const SERVICE_ID = new RegExp(`^svc_${UUID7}$`);
const OPERATION_ID = new RegExp(`^op_${UUID7}$`);
const TIMESTAMP = /^[0-9]{4}-[0-9]{2}-[0-9]{2}T[0-9]{2}:[0-9]{2}:[0-9]{2}\.[0-9]{3}Z$/;
const CONTEXT_FIELDS = [
  "campaign_id", "case_id", "expires_at", "format", "project_id",
  "service_id", "signature",
];
const ED25519_SPKI_PREFIX = Buffer.from("302a300506032b6570032100", "hex");
const ACTIVE_FUZZ_CONTEXT = new AsyncLocalStorage();

export class FuzzContextError extends Error {
  constructor() {
    super("Repro It rejected the fuzz context.");
    this.name = "FuzzContextError";
  }
}

export class FuzzContextValidator {
  #projectId;
  #verificationKey;

  constructor({ projectId, verificationKey }) {
    this.#projectId = projectId;
    this.#verificationKey = verificationKey;
  }

  validate(encoded, now) {
    try {
      if (typeof encoded !== "string" || encoded.length === 0 || encoded.length > 5_462) {
        throw new FuzzContextError();
      }
      if (!/^[A-Za-z0-9_-]+$/.test(encoded) || !/^[A-Za-z0-9_-]+$/.test(this.#verificationKey)) {
        throw new FuzzContextError();
      }
      const raw = Buffer.from(encoded, "base64url");
      const key = Buffer.from(this.#verificationKey, "base64url");
      const value = JSON.parse(raw.toString("utf8"));
      const signature = Buffer.from(value.signature, "base64url");
      if (
        raw.length > MAX_CONTEXT_BYTES || key.length !== 32 || signature.length !== 64 ||
        !sameBytes(canonicalBytes(value), raw) || !sameKeys(value, CONTEXT_FIELDS) ||
        value.format !== "reproit.fuzz-context.v1" ||
        !CAMPAIGN_ID.test(value.campaign_id) || !CASE_ID.test(value.case_id) ||
        !PROJECT_ID.test(value.project_id) || !SERVICE_ID.test(value.service_id) ||
        value.project_id !== this.#projectId ||
        !TIMESTAMP.test(now) || !TIMESTAMP.test(value.expires_at) ||
        !Number.isFinite(Date.parse(now)) || !Number.isFinite(Date.parse(value.expires_at)) ||
        Date.parse(now) >= Date.parse(value.expires_at)
      ) {
        throw new FuzzContextError();
      }
      const unsigned = { ...value, signature: "" };
      const publicKey = createPublicKey({
        format: "der",
        key: Buffer.concat([ED25519_SPKI_PREFIX, key]),
        type: "spki",
      });
      if (!verify(null, canonicalBytes(unsigned), publicKey, signature)) {
        throw new FuzzContextError();
      }
      return Object.freeze({
        campaignId: value.campaign_id,
        caseId: value.case_id,
        contextDigest: `sha256:${createHash("sha256").update(raw).digest("hex")}`,
        encoded,
        now,
        parentOperationId: null,
        projectId: value.project_id,
        serviceId: value.service_id,
        verificationKey: this.#verificationKey,
      });
    } catch (error) {
      if (error instanceof FuzzContextError) throw error;
      throw new FuzzContextError();
    }
  }
}

export function extractHttpFuzzContext(headers, validator, now) {
  const encoded = header(headers, FUZZ_CONTEXT_HTTP_HEADER);
  const parent = header(headers, FUZZ_PARENT_HTTP_HEADER);
  if (encoded === null) {
    if (parent !== null) throw new FuzzContextError();
    return null;
  }
  const context = validator.validate(encoded, now);
  return parent === null ? context : withFuzzParent(context, parent);
}

export function extractQueueFuzzContext(metadata, validator, now) {
  return extractHttpFuzzContext({
    [FUZZ_CONTEXT_HTTP_HEADER]: metadata[FUZZ_CONTEXT_QUEUE_METADATA],
    [FUZZ_PARENT_HTTP_HEADER]: metadata[FUZZ_PARENT_QUEUE_METADATA],
  }, validator, now);
}

export function propagateQueueFuzzContext(metadata) {
  const context = currentFuzzContext();
  if (context === null || metadata === null || typeof metadata !== "object") return;
  metadata[FUZZ_CONTEXT_QUEUE_METADATA] = context.encoded;
  if (context.parentOperationId !== null) {
    metadata[FUZZ_PARENT_QUEUE_METADATA] = context.parentOperationId;
  }
}

export function currentFuzzContext() {
  return ACTIVE_FUZZ_CONTEXT.getStore() ?? null;
}

export function runWithFuzzContext(context, operation) {
  return ACTIVE_FUZZ_CONTEXT.run(context, operation);
}

export function withFuzzParent(context, operationId) {
  if (!OPERATION_ID.test(operationId)) throw new FuzzContextError();
  return Object.freeze({ ...context, parentOperationId: operationId });
}

export function fuzzBeginIdentity(context) {
  return {
    campaign_id: context.campaignId,
    case_id: context.caseId,
    context_digest: context.contextDigest,
  };
}

export function nativeFuzzInput(context) {
  return {
    encoded: context.encoded,
    now: context.now,
    project_id: context.projectId,
    service_id: context.serviceId,
    verification_key: context.verificationKey,
  };
}

function header(headers, expectedName) {
  if (headers === null || typeof headers !== "object") throw new FuzzContextError();
  const selected = Object.entries(headers)
    .filter(([name, value]) => name.toLowerCase() === expectedName.toLowerCase() && value !== undefined)
    .map(([, value]) => value);
  if (selected.length === 0) return null;
  if (selected.length !== 1 || typeof selected[0] !== "string" || selected[0].length === 0) {
    throw new FuzzContextError();
  }
  return selected[0];
}

function sameBytes(left, right) {
  return left.length === right.length && left.equals(right);
}

function sameKeys(value, expected) {
  return value !== null && !Array.isArray(value) && typeof value === "object" &&
    JSON.stringify(Object.keys(value).sort()) === JSON.stringify([...expected].sort());
}

File: sdks/python/reproit_sdk/distributed_fuzz.py
"""Validate and propagate bounded distributed fuzz context."""

from __future__ import annotations

import base64
import contextvars
import datetime
import hashlib
import json
import re
from collections.abc import Mapping, MutableMapping
from dataclasses import dataclass, replace

FUZZ_CONTEXT_HTTP_HEADER = "ReproIt-Fuzz-Context"
FUZZ_PARENT_HTTP_HEADER = "ReproIt-Parent-Operation"
FUZZ_CONTEXT_QUEUE_METADATA = "reproit.fuzz.context"
FUZZ_PARENT_QUEUE_METADATA = "reproit.parent.operation"
_MAX_CONTEXT_BYTES = 4_096
_UUID7 = r"[0-9a-f]{8}-[0-9a-f]{4}-7[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}"
_CAMPAIGN_ID = re.compile(rf"fc_{_UUID7}\Z")
_CASE_ID = re.compile(rf"case_{_UUID7}\Z")
_PROJECT_ID = re.compile(rf"prj_{_UUID7}\Z")
_SERVICE_ID = re.compile(rf"svc_{_UUID7}\Z")
_OPERATION_ID = re.compile(rf"op_{_UUID7}\Z")
_TIMESTAMP = re.compile(r"[0-9]{4}-[0-9]{2}-[0-9]{2}T[0-9]{2}:[0-9]{2}:[0-9]{2}\.[0-9]{3}Z\Z")
_FIELDS = frozenset(
    {
        "campaign_id",
        "case_id",
        "expires_at",
        "format",
        "project_id",
        "service_id",
        "signature",
    }
)


class FuzzContextError(ValueError):
    """The fuzz context failed bounded validation."""


@dataclass(frozen=True)
class FuzzCampaignContext:
    """A validated context and its native signature-verification input."""

    campaign_id: str
    case_id: str
    context_digest: str
    encoded: str
    now: str
    parent_operation_id: str | None
    project_id: str
    service_id: str
    verification_key: str

    def with_parent(self, operation_id: str) -> FuzzCampaignContext:
        """Return the context for a causal child operation."""
        if not _OPERATION_ID.fullmatch(operation_id):
            raise FuzzContextError("The parent operation ID is invalid.")
        return replace(self, parent_operation_id=operation_id)

    def native_input(self) -> dict[str, str]:
        """Return the bounded input that the shared engine verifies."""
        return {
            "encoded": self.encoded,
            "now": self.now,
            "project_id": self.project_id,
            "service_id": self.service_id,
            "verification_key": self.verification_key,
        }

    def begin_identity(self) -> dict[str, str]:
        """Return the bounded identity stored in OperationBegin v2."""
        return {
            "campaign_id": self.campaign_id,
            "case_id": self.case_id,
            "context_digest": self.context_digest,
        }


@dataclass(frozen=True)
class FuzzContextValidator:
    """Validate context structure before native signature verification."""

    project_id: str
    verification_key: str

    def validate(self, encoded: str, now: str) -> FuzzCampaignContext:
        """Validate one opaque context and prepare native verification."""
        try:
            raw = _decode_base64url(encoded)
            key = _decode_base64url(self.verification_key)
            value = json.loads(raw, object_pairs_hook=_unique_object)
            canonical = _canonical_bytes(value)
            signature = _decode_base64url(value["signature"])
            if not _TIMESTAMP.fullmatch(now) or not _TIMESTAMP.fullmatch(
                value["expires_at"]
            ):
                raise ValueError
            current = datetime.datetime.strptime(
                now,
                "%Y-%m-%dT%H:%M:%S.%fZ",
            ).replace(tzinfo=datetime.UTC)
            expires = datetime.datetime.strptime(
                value["expires_at"],
                "%Y-%m-%dT%H:%M:%S.%fZ",
            ).replace(tzinfo=datetime.UTC)
        except (KeyError, TypeError, ValueError, UnicodeError) as error:
            raise FuzzContextError("The fuzz context is invalid.") from error
        if (
            len(raw) > _MAX_CONTEXT_BYTES
            or len(key) != 32
            or len(signature) != 64
            or canonical != raw
            or set(value) != _FIELDS
            or value["format"] != "reproit.fuzz-context.v1"
            or not _CAMPAIGN_ID.fullmatch(value["campaign_id"])
            or not _CASE_ID.fullmatch(value["case_id"])
            or not _PROJECT_ID.fullmatch(value["project_id"])
            or not _SERVICE_ID.fullmatch(value["service_id"])
            or value["project_id"] != self.project_id
            or current >= expires
        ):
            raise FuzzContextError("The fuzz context is invalid.")
        return FuzzCampaignContext(
            campaign_id=value["campaign_id"],
            case_id=value["case_id"],
            context_digest="sha256:" + hashlib.sha256(raw).hexdigest(),
            encoded=encoded,
            now=now,
            parent_operation_id=None,
            project_id=value["project_id"],
            service_id=value["service_id"],
            verification_key=self.verification_key,
        )


_ACTIVE_FUZZ_CONTEXT: contextvars.ContextVar[FuzzCampaignContext | None] = (
    contextvars.ContextVar("reproit_fuzz_context", default=None)
)


def extract_http_fuzz_context(
    headers: Mapping[str, str],
    validator: FuzzContextValidator,
    now: str,
) -> FuzzCampaignContext | None:
    """Validate inbound HTTP metadata without starting an operation."""
    encoded = _header(headers, FUZZ_CONTEXT_HTTP_HEADER)
    parent = _header(headers, FUZZ_PARENT_HTTP_HEADER)
    if encoded is None:
        if parent is not None:
            raise FuzzContextError("The fuzz parent has no fuzz context.")
        return None
    context = validator.validate(encoded, now)
    if parent is None:
        return context
    return context.with_parent(parent)


def extract_queue_fuzz_context(
    metadata: Mapping[str, str],
    validator: FuzzContextValidator,
    now: str,
) -> FuzzCampaignContext | None:
    """Validate inbound delivered-work metadata."""
    translated: dict[str, str] = {}
    if FUZZ_CONTEXT_QUEUE_METADATA in metadata:
        translated[FUZZ_CONTEXT_HTTP_HEADER] = metadata[FUZZ_CONTEXT_QUEUE_METADATA]
    if FUZZ_PARENT_QUEUE_METADATA in metadata:
        translated[FUZZ_PARENT_HTTP_HEADER] = metadata[FUZZ_PARENT_QUEUE_METADATA]
    return extract_http_fuzz_context(translated, validator, now)


def propagate_queue_fuzz_context(metadata: MutableMapping[str, str]) -> None:
    """Add the active context to outbound delivered-work metadata."""
    context = _ACTIVE_FUZZ_CONTEXT.get()
    if context is None:
        return
    metadata[FUZZ_CONTEXT_QUEUE_METADATA] = context.encoded
    if context.parent_operation_id is not None:
        metadata[FUZZ_PARENT_QUEUE_METADATA] = context.parent_operation_id


def _current_fuzz_context() -> FuzzCampaignContext | None:
    return _ACTIVE_FUZZ_CONTEXT.get()


def _activate_fuzz_context(
    context: FuzzCampaignContext | None,
) -> contextvars.Token[FuzzCampaignContext | None]:
    return _ACTIVE_FUZZ_CONTEXT.set(context)


def _reset_fuzz_context(
    token: contextvars.Token[FuzzCampaignContext | None],
) -> None:
    _ACTIVE_FUZZ_CONTEXT.reset(token)


def _header(headers: Mapping[str, str], name: str) -> str | None:
    selected = [value for key, value in headers.items() if key.lower() == name.lower()]
    if not selected:
        return None
    if len(selected) != 1 or not isinstance(selected[0], str) or not selected[0]:
        raise FuzzContextError("The fuzz header is invalid.")
    return selected[0]


def _decode_base64url(value: str) -> bytes:
    if not isinstance(value, str) or not value or len(value) > 5_462:
        raise ValueError
    if re.fullmatch(r"[A-Za-z0-9_-]+", value) is None:
        raise ValueError
    return base64.urlsafe_b64decode(value + "=" * (-len(value) % 4))


def _unique_object(pairs: list[tuple[str, object]]) -> dict[str, object]:
    value: dict[str, object] = {}
    for key, element in pairs:
        if key in value:
            raise ValueError
        value[key] = element
    return value


def _canonical_bytes(value: object) -> bytes:
    return json.dumps(
        value,
        ensure_ascii=False,
        separators=(",", ":"),
        sort_keys=True,
    ).encode("utf-8", "strict")
